File: typescipt/src/compiler/JavaScript/writers/functionWriter.ts
import type {IRootTokenWriter} from "../../IRootTokenWriter";
import type {IRootNode} from "../../../language/rootNode";
import {LexyCodeConstants} from "../../lexyCodeConstants";
import {GeneratedType, GeneratedTypeKind} from "../../generatedType";
import {asFunction, Function} from "../../../language/functions/function";
import {NodesWalker} from "../../../language/nodesWalker";
import {FunctionCall} from "../builtInFunctions/functionCall";
import {asFunctionCallExpression} from "../../../language/expressions/functionCallExpression";
import {CodeWriter} from "./codeWriter";
import {functionClassName} from "../classNames";
import {renderExpressions} from "./renderExpression";
import {createVariableClass} from "./renderVariableClass";
import {createFunctionCall} from "../builtInFunctions/createFunctionCall";

export class FunctionWriter implements IRootTokenWriter {

  private readonly namespace: string;

  constructor(namespace: string) {
    this.namespace = namespace;
  }

  public createCode(node: IRootNode): GeneratedType {
    const functionNode = asFunction(node);
    if (functionNode == null) throw new Error(`Root token not Function`);

    let builtInFunctionCalls = this.getBuiltInFunctionCalls(functionNode);
    const codeWriter = new CodeWriter(this.namespace, builtInFunctionCalls)

    return this.createFunction(functionNode, codeWriter);
  }

  private createFunction(functionNode: Function, codeWriter: CodeWriter) {

    codeWriter.openScope("function scope()");
    this.renderCustomBuiltInFunctions(functionNode, codeWriter)
    this.renderRunFunction(functionNode, codeWriter);
    createVariableClass(LexyCodeConstants.parametersType, functionNode.parameters.variables, codeWriter);
    createVariableClass(LexyCodeConstants.resultsType, functionNode.results.variables, codeWriter);
    codeWriter.writeLine(`${LexyCodeConstants.runMethod}.${LexyCodeConstants.parametersType} = ${LexyCodeConstants.parametersType};`)
    codeWriter.writeLine(`${LexyCodeConstants.runMethod}.${LexyCodeConstants.resultsType} = ${LexyCodeConstants.resultsType};`)
    codeWriter.writeLine(`return ${LexyCodeConstants.runMethod};`)
    codeWriter.closeScope("();");

    return new GeneratedType(GeneratedTypeKind.function, functionNode, functionClassName(functionNode.nodeName), codeWriter.toString());
  }

  private renderRunFunction(functionNode: Function, codeWriter: CodeWriter) {
    
    codeWriter.openScope(`function ${LexyCodeConstants.runMethod}(${LexyCodeConstants.parameterVariable}, ${LexyCodeConstants.contextVariable})`);

    this.renderResults(functionNode, codeWriter)
    this.renderCode(functionNode, codeWriter);

    codeWriter.writeLine(`return ${LexyCodeConstants.resultsVariable};`);

    codeWriter.closeScope();
  }


  private getBuiltInFunctionCalls(functionNode: Function): Array<FunctionCall> {
    return NodesWalker.walkWithResult(functionNode.code.expressions,
      node => {
        const expression = asFunctionCallExpression(node);
        if (expression != null) {
          return createFunctionCall(expression)
        } else {
          return null
        }
      });
  }

  private renderCode(functionNode: Function, codeWriter: CodeWriter) {
    renderExpressions(functionNode.code.expressions, codeWriter);
  }

  private renderResults(functionNode: Function, codeWriter: CodeWriter) {
    codeWriter.writeLine(`const ${LexyCodeConstants.resultsVariable} = new ${LexyCodeConstants.resultsType}();`);
  }

  private renderCustomBuiltInFunctions(functionNode: Function, codeWriter: CodeWriter) {
    codeWriter.renderCustomBuiltInFunctions();
  }
}

File: src/Lexy.Compiler/Compiler/CSharp/Writers/FunctionWriter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Lexy.Compiler.Compiler.CSharp.BuiltInFunctions;
using Lexy.Compiler.Compiler.CSharp.Syntax;
using Lexy.Compiler.Language;
using Lexy.Compiler.Language.Expressions;
using Lexy.Compiler.Language.Functions;
using Lexy.Compiler.Language.Scenarios;
using Lexy.Compiler.Language.VariableTypes;
using Lexy.RunTime;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
using static Lexy.Compiler.Compiler.CSharp.Syntax.Expressions;

namespace Lexy.Compiler.Compiler.CSharp.Writers;

public class FunctionWriter : IRootTokenWriter
{
    public GeneratedClass CreateCode(IRootNode node)
    {
        if (node is not Function function) throw new InvalidOperationException("Root token not Function");

        var members = new List<MemberDeclarationSyntax>
        {
            VariableClass.TranslateVariablesClass(LexyCodeConstants.ParametersType, function.Parameters.Variables),
            VariableClass.TranslateVariablesClass(LexyCodeConstants.ResultsType, function.Results.Variables),
            RunMethod(function)
        };
        members.AddRange(CustomBuiltInFunctions(function));

        var name = ClassNames.FunctionClassName(node.NodeName);

        var classDeclaration = ClassDeclaration(name)
            .WithModifiers(Modifiers.PublicStatic())
            .WithMembers(List(members));

        return new GeneratedClass(function, name, classDeclaration);
    }

    private IEnumerable<MemberDeclarationSyntax> CustomBuiltInFunctions(Function function)
    {
        return NodesWalker.WalkWithResult(function.Code.Expressions,
            node => node is FunctionCallExpression expression ? FunctionCallFactory.CustomMethods(expression.ExpressionFunction) : null)
            .Where(value => value != null);
    }

    private MethodDeclarationSyntax RunMethod(Function function)
    {
        var statements = new List<StatementSyntax>
        {
            GuardStatements.VerifyNotNull(LexyCodeConstants.ParameterVariable),
            GuardStatements.VerifyNotNull(LexyCodeConstants.ContextVariable),
            LogCalls.SetFileName(function.Reference.File.FileName),
            LogCalls.OpenScope($"Execute: {function.NodeName}", function.Reference.LineNumber ?? -1),
        };

        if (function.Parameters != null)
        {
            statements.Add(LogCalls.LogVariables(function.Parameters.Reference?.LineNumber, "Parameters",
                LexyCodeConstants.ParameterVariable));
        }

        statements.Add(InitializeResults());

        statements.AddRange(ExecuteExpressionStatementSyntax(function.Code.Expressions, false));
        if (function.Results != null)
        {
            statements.Add(LogCalls.LogVariables(function.Results.Reference.LineNumber, "Results",
                LexyCodeConstants.ResultsVariable));
        }

        statements.Add(LogCalls.CloseScope());
        statements.Add(ReturnResults());

        var functionSyntax = MethodDeclaration(
                IdentifierName(LexyCodeConstants.ResultsType),
                Identifier(LexyCodeConstants.RunMethod))
            .WithModifiers(Modifiers.PublicStatic())
            .WithParameterList(
                ParameterList(
                    SeparatedList<ParameterSyntax>(
                        new SyntaxNodeOrToken[]
                        {
                            Parameter(Identifier(LexyCodeConstants.ParameterVariable))
                                .WithType(IdentifierName(LexyCodeConstants.ParametersType)),
                            Token(SyntaxKind.CommaToken),
                            Parameter(Identifier(LexyCodeConstants.ContextVariable))
                                .WithType(IdentifierName(nameof(IExecutionContext)))
                        })))
            .WithBody(Block(statements));

        return functionSyntax;
    }

    private IEnumerable<VariableUsage> Variables(ComplexType complexType, IReadOnlyList<VariableDefinition> variables, VariableSource source)
    {
        return variables.Select(variable => new VariableUsage(
            VariablePathParser.Parse(variable.Name),
            complexType,
            variable.VariableType,
            source,
            VariableAccess.Read));
    }

    private StatementSyntax ReturnResults()
    {
        return ReturnStatement(IdentifierName(LexyCodeConstants.ResultsVariable));
    }

    private StatementSyntax InitializeResults()
    {
        return LocalDeclarationStatement(
            VariableDeclaration(
                    IdentifierName(
                        Identifier(TriviaList(), SyntaxKind.VarKeyword, "var", "var", TriviaList())))
                .WithVariables(
                    SingletonSeparatedList(
                        VariableDeclarator(
                                Identifier(LexyCodeConstants.ResultsVariable))
                            .WithInitializer(
                                EqualsValueClause(
                                    ObjectCreationExpression(
                                            IdentifierName(LexyCodeConstants.ResultsType))
                                        .WithArgumentList(
                                            ArgumentList()))))));
    }
}